db/index_changes.go

```diff
@@ -403,17 +422,105 @@ func (db *Database) initializeChannelFeeds(channelsSince channels.TimedSet, opti
+		}
+		fromClock.SetSequence(vbNo, fromSeq)
+
+		// Calculate to sequence for vbucket
+		if vbNo < triggerPosition.Vb {
+			toSeq = MAX_SEQUENCE
+		} else if vbNo > triggerPosition.Vb {
+			toSeq = sinceClock.GetSequence(vbNo)
+		} else { // vbNo==triggerPosition.Vb
+			toSeq = triggerPosition.Seq
+		}
+		toClock.SetSequence(vbNo, toSeq)
+	}
+
+	return fromClock, toClock
+}
+
 // Creates a Go-channel of all the changes made on a channel.
 // Does NOT handle the Wait option. Does NOT check authorization.
 func (db *Database) vectorChangesFeed(channel string, options ChangesOptions) (<-chan *ChangeEntry, error) {
 	dbExpvars.Add("channelChangesFeeds", 1)
-	log, err := db.changeCache.GetChanges(channel, options)
-	base.LogTo("Changes+", "[changesFeed] Found %d changes for channel %s", len(log), channel)
-	if err != nil {
-		return nil, err
+	changeIndex, ok := db.changeCache.(*kvChangeIndex)
+	if !ok {
+		return nil, errors.New("Called vectorChangesFeed with non-index cache.")
 	}
 
-	if len(log) == 0 {
+	// If we're in backfill for this channel, we make one reader call for the backfill and one for non-backfill.  Without
+	// two requests, it's not possible to use limit to prevent a full index scan of some vbuckets while still preserving ordering.
+	// The second call should only be made if the first request doesn't return limit changes.
+	var backfillLog, log []*LogEntry
+	var err error
+
+	if options.Since.TriggeredByClock != nil {
+		// Changes feed is in backfill for this channel.
+		// Backfill position: (vb,seq) position in the backfill. e.g. [0,0] if we're just starting the backfill, [vb,seq] if we're midway through.
+		// Trigger position: (vb,seq) of the document that triggered this backfill (e.g. access granting doc or user doc)
+		backfillPosition := base.VbSequence{options.Since.vbNo, options.Since.Seq}
+		triggerPosition := base.VbSequence{options.Since.TriggeredByVbNo, options.Since.TriggeredBy}
+		backfillFrom, backfillTo := calculateBackfillRange(backfillPosition, triggerPosition, options.Since.Clock)
+
+		backfillLog, err = changeIndex.reader.GetChangesForRange(channel, backfillFrom, backfillTo, options.Limit)
+		if err != nil {
+			return nil, err
+		}
+		base.LogTo("Changes+", "[changesFeed] Found %d backfill changes for channel %s", len(backfillLog), channel)
+
+		// If we still have room, get non-backfill entries
+		if options.Limit == 0 || len(backfillLog) < options.Limit {
+			log, err = changeIndex.reader.GetChangesForRange(channel, backfillTo, nil, options.Limit)
+			if err != nil {
+				return nil, err
+			}
+			base.LogTo("Changes+", "[changesFeed] Found %d non-backfill changes for channel %s", len(log), channel)
+		}
+
+	} else {
+		// Not backfill for this channel.  Standard changes processing
+		log, err = changeIndex.reader.GetChangesForRange(channel, options.Since.Clock, nil, options.Limit)
+		base.LogTo("Changes+", "[changesFeed] Found %d changes for channel %s", len(log), channel)
+		if err != nil {
```

Comment:
Could the `base.LogTo` come after the err check?  Otherwise I think if it's before, it should log the error to.  

I mean we should have either:

```
log, err = changeIndex.reader.GetChangesForRange(channel, options.Since.Clock, nil, options.Limit)
base.LogTo("Changes+", "[changesFeed] Found %d changes for channel %s.  Err: %v", len(log), channel, err)
if err != nil {
```

or 

```
log, err = changeIndex.reader.GetChangesForRange(channel, options.Since.Clock, nil, options.Limit)
if err != nil {
                  ...
}
base.LogTo("Changes+", "[changesFeed] Found %d changes for channel %s", len(log), channel)
```